добавлена логика лайков и избранных + рецепт дня

tanyginvv opened the pull request.

tanyginvv committed:
добавлена логика лайков и избранных+ рецепт дня

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -105,15 +108,19 @@ public async Task<ActionResult<RecipeReadDto>> GetRecipeById(
         [FromQuery] int pageNumber = 1,
-        [FromQuery] List<string> searchTerms = null )
+        [FromQuery] List<string> searchTerms = null,
+        [FromQuery] bool isAuth = false,
```
странное название для флага, давай тут енамчик добавим лучше, это же взаимозаменяемые по-сути (нельзя одновременно искать и свои и избранные) , поэтому можно ввести enum вида 
RecipeSelection
{
  All,
  My,
  Starred
}

Или вот чатжпт предлагает, можно его послушать
public enum RecipeQueryType
{
    All,
    My,
    Starred
}

VladDotMe commented on backend/Recipes/Recipes.Application/Filters/UserFilter.cs:
```diff
@@ -1,19 +1,29 @@
 public class UserFilter : IFilter<Recipe>
 {
     public int UserId { get; set; }
+    public bool IsFavourite { get; set; }
```
Тут тоже енамчик применить, на слой Application и смаппить из WebApi

VladDotMe commented on backend/Recipes/Recipes.Application/Filters/UserFilter.cs:
```diff
@@ -1,19 +1,29 @@
     public IQueryable<Recipe> Apply( IQueryable<Recipe> query )
     {
-        if ( UserId != 0 )
+        if ( UserId != 0 && !IsFavourite && IsAuth )
```
И тут проверять на енам, если My то возвращать по автору, если Starred то избранные, если UserId == 0 / All то можно сразу возвращать изначальное query (guard clause)

VladDotMe commented on backend/Recipes/Recipes.Domain/Entities/Like.cs:
```diff
@@ -0,0 +1,15 @@
+﻿namespace Recipes.Domain.Entities;
+
+public class Like : Entity
```
Чётко сделал связи!

VladDotMe commented on backend/Recipes/Recipes.Domain/Entities/User.cs:
```diff
@@ -7,6 +7,8 @@ public class User : Entity
     public string Login { get; set; }
     public string PasswordHash { get; set; }
     public IReadOnlyList<Recipe> Recipes { get; private set; }
+    public ICollection<Like> Likes { get; set; }
```
У Recipes выше может тоже ICollection тогда?

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Queries/GetRecipes/GetRecipesQueryHandler.cs:
```diff
@@ -10,22 +10,35 @@ namespace Recipes.Application.UseCases.Recipes.Queries.GetRecipes;
+            {
+                new SearchFilter { SearchTerms = query.SearchTerms },
+                new UserFilter { UserId = query.UserId, IsFavourite = query.isFavourite, IsAuth = query.IsAuth },
+                new PaginationFilter { PageNumber = query.PageNumber, PageSize = 4 },
```
PageSize в константы, можно вида PaginationFilter.DefaultPageSize

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Queries/GetRecipes/GetRecipesQueryHandler.cs:
```diff
@@ -10,22 +10,35 @@ namespace Recipes.Application.UseCases.Recipes.Queries.GetRecipes;
 
+        if ( query.UserId != 0 )
+        {
+            foreach ( GetRecipePartDto recipeDto in recipeDtos )
```
у тебя же все данные есть, зачем снова ходить в базу? да еще и по каждому рецепту.. напиши мапу просто и оттуда бери

типа такого

            List<int> likedRecipes = recipes.Where( r => r.Likes.Any( l => l.UserId == query.UserId ) ).Select( r => r.Id ).ToList();
            List<int> starredRecipes = recipes.Where( r => r.Favourites.Any( l => l.UserId == query.UserId ) ).Select( r => r.Id ).ToList();

            foreach ( GetRecipePartDto recipeDto in recipeDtos )
            {
                if ( likedRecipes.Contains( recipeDto.Id ) )
                {
                    recipeDto.IsLike = true;
                }

                if ( starredRecipes.Contains( recipeDto.Id ) )
                {
                    recipeDto.IsFavourite = true;
                }
            }

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/DataAccess/Likes/LikeRepository.cs:
```diff
@@ -0,0 +1,36 @@
+
+    public async Task<Like> GetLikeByAttributes( int userId, int recipeId )
+    {
+        return await _dbSet
```
Во-первых, Where + FirstOrDefaultAsync можно объединить в одну команду. А во-вторых, что думаешь насчет того, чтобы добавить сюда индекс? И какой?)

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Likes/LikeBinding.cs:
```diff
@@ -0,0 +1,21 @@
+    {
+        services.AddScoped<ICommandHandler<CreateLikeCommand>, CreateLikeCommandHandler>();
+        services.AddScoped<ICommandHandler<DeleteLikeCommand>, DeleteLikeCommandHandler>();
+
```
две строки, чет перебор.

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/DataAccess/Users/UserConfiguration.cs:
```diff
@@ -38,5 +38,15 @@ public void Configure( EntityTypeBuilder<User> builder )
+        builder.HasMany( u => u.Likes )
+            .WithOne( ua => ua.User )
+            .HasForeignKey( ua => ua.UserId )
+            .OnDelete( DeleteBehavior.Restrict );
```
А почему тут Restrict а выше Cascade? В целом разные сервисы имеют разные политики удаления пользователей, но редко кто дает возможность удалять пользователей полностью. 
Вообщем, Restrict норм, но я бы у Recipes тож Restrict поставил.

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/DataAccess/Recipes/RecipeConfiguration.cs:
```diff
@@ -33,14 +33,24 @@ public void Configure( EntityTypeBuilder<Recipe> builder )
+        builder.HasMany( r => r.Steps )
             .WithOne( s => s.Recipe )
             .HasForeignKey( s => s.RecipeId )
             .OnDelete( DeleteBehavior.Cascade );
```
а вот здесь везде Cascade точно норм)

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Dtos/GetRecipeQueryDto.cs:
```diff
@@ -3,11 +3,16 @@
     public required int CookTime { get; init; }
     public required int PortionCount { get; init; }
+    public required int LikeCount { get; set; }
+    public required bool IsLike { get; set; }
```
IsLiked и IsFavourited

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Dtos/GetRecipePartDto.cs:
```diff
@@ -8,6 +8,10 @@ public class GetRecipePartDto
     public required int CookTime { get; init; }
     public required int PortionCount { get; init; }
+    public required int LikeCount { get; init; }
+    public required bool IsLike { get; set; }
```
тоже самое, IsLiked / IsFavourited

VladDotMe commented:
Не заметил еще в прошлый раз, а зачем в GetRecipeByIdQueryValidator 
        if ( await repository.GetByIdAsync( query.Id ) is null )
        {
            return Result.FromError( "Рецепта с этим Id не существует" );
        }
        
        у тебя же в самом Query потом аналогичный запрос)
        
                Recipe foundRecipe = await recipeRepository.GetByIdAsync( query.Id );
        if ( foundRecipe is null )
        {
            return Result<GetRecipeQueryDto>.FromError( "Рецепт не найден" );
        }

        
        Два раза тоже самое делаешь получается..

VladDotMe reviewed (changes requested):
Пока так, еще вопрос про валидацию у команд, где в базу ходишь. Мне кажется избыточным, особенно у CreateLike, ибо база сама все эти три проверки выполнит при инсерте (если индексы правильно настроил). Но я пока не на 100% уверен как лучше, я завтра тебе скажу, пока можешь подумать тоже)

tanyginvv committed:
правки по комментам

1)Добавил enum для получения рецептов
2)добавил составные индексы для поиска лайков и избранных

tanyginvv committed:
букву не ту поставил

tanyginvv committed:
микро-правка

tanyginvv committed:
обновления с ветки фронтенда

VladDotMe commented on backend/Recipes/Recipes.Application/Filters/UserFilter.cs:
```diff
@@ -1,23 +1,22 @@
+        if ( RecipeQueryType == RecipeQueryType.Starred && UserId > 0 )
         {
             query = query
                 .Where( u => u.Favourites
```
Пару мелочей:  у тебя .Any внутри .Where, так не надо переносить на следующую строчку) 
Либо на этой же либо переносить со сдвигом на 4 пробела вправо (IDE вроде по умолчанию так должна сделать), потому что иначе неочевидно

Ну и UserId сверху проверяешь != 0 а тут > 0, можно подумать, что какая-то разная специфика, но ее нет) Просто приведи к одному виду

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Queries/GetRecipes/GetRecipesQueryHandler.cs:
```diff
@@ -10,35 +10,53 @@ namespace Recipes.Application.UseCases.Recipes.Queries.GetRecipes;
+                new PaginationFilter { PageNumber = query.PageNumber, PageSize = PaginationFilter.DefaultPageSize },
             } );
 
+        IEnumerable<Recipe> recipesNext = await recipeRepository.GetRecipesAsync(
```
За что ты так жесток с базой данных?) Два одинаковых запроса с Include подряд, ухх.. .. И еще считаешь потом элементы.. А не проще сразу базу спросить есть ли что-то еще?) 

    public async Task<List<Recipe>> AnyAsync( IEnumerable<IFilter<Recipe>> filters )
    {
        return await _dbSet
           .ApplyFilters( filters )
           .AnyAsync();
    }
    
   Ну и фильтры лучше переиспользовать а не писать два раза тоже самое. Только один PaginationFilter модифицировать.

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Queries/GetRecipes/GetRecipesQueryHandler.cs:
```diff
@@ -10,35 +10,53 @@ namespace Recipes.Application.UseCases.Recipes.Queries.GetRecipes;
         List<GetRecipePartDto> recipeDtos = recipes.Adapt<List<GetRecipePartDto>>();
 
-        if ( query.UserId != 0 )
+        List<int> likedRecipes = recipes.Where( r => r.Likes.Any( l => l.UserId == query.UserId ) ).Select( r => r.Id ).ToList();
```
Ща еще понял, что здесь больше подходит HashSet вместо List ) Потому что ниже идет поиск if (likedRecipes.Contains(..)), для List это O(N), для HashSet O(1)

VladDotMe commented on backend/Recipes/Recipes.WebApi/Dto/UserDtos/UpdateUserDto.cs:
```diff
@@ -8,17 +8,14 @@ public class UpdateUserDto
     [MaxLength( 50 )]
     public string Login { get; init; }
 
-    [Required]
```
а у паролей зачем убрал Required? Они же обязательные

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/DataAccess/Likes/LikeRepository.cs:
```diff
@@ -0,0 +1,35 @@
+
+public class LikeRepository( RecipesDbContext context ) : BaseRepository<Like>( context ), ILikeRepository
+{
+    public async Task<Like> GetByIdAsync( Like like )
```
Не используется метод (он у тебя в базовом репозитории есть)

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/DataAccess/Favourites/FavouriteRepository.cs:
```diff
@@ -0,0 +1,35 @@
+
+public class FavouriteRepository( RecipesDbContext context ) : BaseRepository<Favourite>( context ), IFavouriteRepository
+{
+    public async Task<Favourite> GetByIdAsync( Favourite fav )
```
этот тоже можно убрать, использовать базовый GetByIdAsync

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/DataAccess/Favourites/FavouriteRepository.cs:
```diff
@@ -0,0 +1,35 @@
+        }
+    }
+
+    public async Task<Favourite> GetFavouriteByAttributes( int recipeId, int userId )
```
А этот дублируется с верхним ContainsAsync. Почему иногда используешь этот, иногда тот?
Ну и еще если ты индекс добавил на RecipeId, UserId то тебе надо искать в базе именно в таком порядке. (но правда чатжпт говорит, что SQL Server умная и сама аргументы поменяет местами под индекс, но в целом все равно правильно писать запрос в том же порядке как у тебя индекс построен)

Я бы на твоем месте чтобы это гарантировать, убрал бы верхний ContainsAsync и оставил бы только GetByRecipeIdAndUserId. 

Все аналогичные комменты относятся и к LikeRepository

tanyginvv committed:
правки по комментам

tanyginvv committed:
правки по атрибутам лайков и избранных

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Users/Commands/UpdateUser/UpdateUserCommandHandler.cs:
```diff
@@ -21,8 +21,16 @@ protected override async Task<Result> HandleImplAsync( UpdateUserCommand command
+            user.Name = command.Name;
+        }
+
+        if ( command.Description is not null )
```
тут может тоже !string.IsNullOrEmpty ?
Хотя лучше везде вообще использовать string.IsNullOrWhiteSpace

Чтобы строки с миллионом пробелов не пролазили

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Likes/Command/CreateLike/CreateLikeCommandValidator.cs:
```diff
@@ -22,7 +22,7 @@ public async Task<Result> ValidateAsync( CreateLikeCommand command )
         }
 
-        if ( await likeRepository.ContainsAsync( u => u.UserId == command.UserId && u.RecipeId == command.RecipeId ) )
+        if ( await likeRepository.GetLikeByAttributes( command.UserId, command.RecipeId ) is not null )
```
Здесь не в том порядке передаешь (в recipeId передаешь юзера и наоборот)

tanyginvv committed:
last правки
